### DISCUSSION TITLE
How should I use the types and constants in the Qt Namespace?

### DISCUSSION THREAD

**Discussion by ghost** [KDAB/cxx-qt#890] [Q&A] [answered]
I am trying to create a TableView model. The stumbling block here is the handling of what is used in C++ in the form of Qt::~.
For example, Qt::Orientation used as a headerData argument is not a type in cxx_qt_lib, but I don't know how to use it. I tried to create an enum called Orientation with reference to the sample custom_base_class.rs Role, but I get an error saying `cannot convert 'Orientation' to 'Qt::Orientation'`.
Or something like Qt::DisplayRole, I know it can be represented as a number, but is there a way to use it as a constant?
Sorry if this is an oversight or misunderstanding, but I would appreciate your help.

**Comment by LeonMatthesKDAB** [comment]
Hi, thank you for taking the time to ask this question :)

What you're probably looking for is an `extern enum` as described here: https://cxx.rs/shared.html#extern-enums

So for Qt::Orientation, you likely want something like this:
```rust
#[cxx_qt::bridge(namespace="Qt")]
mod qt {
    enum Orientation {
        Horizontal,
        Vertical,
    }

    unsafe extern "C++" {
        include!(< Qt >);
        type  Orientation; // <---- this tells CXX that the enum shouldn't be a new type, but already exists in C++.
    }
}
```

Then you should be able to use `qt::Orientation` for the `headerData` argument.
Hope that helps :four_leaf_clover:

**Reply by ghost** [reply]
Thank you for your reply.
I am sorry, but I am not able to get it to work.
It seems that only one cxx_qt object can be placed in a single file, so I tried creating a file called qt.rs, for example, and writing it there, but I get the error 

  ```
error[cxx]: unsupported type
     ┌─ /home/username/projects/projectname/src/qml_objects/table.rs:45:26
     │
  45 │             orientation: qt::Orientation,
     │                          ^^^^^^^^^^^^^^^
```
* I use it after `use crate::qml_objects::qt::qt;`.

On the other hand, if I write it in the same module where I wrote headerData(), this error does not occur, but it fails because I cannot specify the namespace.
What am I missing?

**Reply by LeonMatthesKDAB** [reply]
No problem.

If you have the `Orientation` enum in another file, you have to import it, so that CXX knows about it.
See: https://cxx.rs/extern-c++.html#reusing-existing-binding-types

Alternatively, you should be able to change the namespace of the enum only.
I think it is done like this:
```rust
#[cxx_qt::bridge]
mod qt {
    enum Orientation {
        Horizontal,
        Vertical,
    }

    unsafe extern "C++" {
        include!(< Qt >);
        #[namespace="Qt"] // <---- only set the namespace for the one type
        type  Orientation;
    }
}
```

**Reply by ghost** [reply]
Thank you very much.
I think we have made progress, but it doesn't seem to be working.
I have tried both methods and get the following error.(Excerpts)

```
warning: projectname@0.1.0: 次のファイルから読み込み:  /home/username/projects/projectname/target/debug/build/projectname-99c3881ccd5a0cce/out/cxx-qt-gen/qobject.cxxqt.h:11,
warning: projectname@0.1.0:          次から読み込み:  /home/username/projects/projectname/target/debug/build/projectname-99c3881ccd5a0cce/out/cxx-qt-gen/src/qobject.cxx.cpp:8:
warning: projectname@0.1.0: /home/username/projects/projectname/target/debug/build/projectname-99c3881ccd5a0cce/out/cxx-qt-gen/qobject.cxx.h:293:30: エラー: ‘Orientation’ was not declared in this scope; did you mean ‘Qt::Orientation’?
warning: projectname@0.1.0:   293 | static_assert(::std::is_enum<Orientation>::value, "expected enum");
warning: projectname@0.1.0:       |                              ^~~~~~~~~~~
warning: projectname@0.1.0:       |                              Qt::Orientation
warning: projectname@0.1.0: /home/username/projects/projectname/target/debug/build/projectname-99c3881ccd5a0cce/out/cxx-qt-gen/qobject.cxx.h:287:9: 備考: ‘Qt::Orientation’ declared here
warning: projectname@0.1.0:   287 |   using Orientation = ::Qt::Orientation;
warning: projectname@0.1.0:       |         ^~~~~~~~~~~
warning: projectname@0.1.0: /home/username/projects/projectname/target/debug/build/projectname-99c3881ccd5a0cce/out/cxx-qt-gen/qobject.cxx.h:293:41: エラー: template argument 1 is invalid
```
※The Japanese has been mixed up in places, but it means the following.
エラー：Error
次のファイルから読み込み：Read from the following file
次から読み込み:Read from following
備考:Remarks

I thought using was to define an alias, so I am not sure what this error means either. Sorry for my lack of study, but what am I doing wrong?

**Reply by LeonMatthesKDAB** [reply]
Ah, sorry, I forgot you need the `#[namespace=""]` attribute both in the `extern "C++"`  and on the `enum Orientation`.

So should work like this:
```rust
#[cxx_qt::bridge]
mod qt {
    #[namespace="Qt"] // <--- here
    enum Orientation {
        Horizontal,
        Vertical,
    }

    unsafe extern "C++" {
        include!(< Qt >);
        #[namespace="Qt"] // <---- and here
        type  Orientation;
    }
}
```

Hope that works :crossed_fingers:

**Reply by ghost** [reply]
Thank you very much.
Thanks to your help, the namespace problem is solved, and next I am suffering from static assertion failure.
The error is as follows

```
  cargo:warning=  295 | static_assert(sizeof(Orientation) == sizeof(::std::uint8_t), "incorrect size");

  cargo:warning=      |               ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~

  cargo:warning=/home/username/projects/projectname/target/debug/build/projectname-99c3881ccd5a0cce/out/cxx-qt-gen/qobject.cxx.h:295:35: 備考: the comparison reduces to ‘(4 == 1)’

  cargo:warning=/home/username/projects/projectname/target/debug/build/projectname-99c3881ccd5a0cce/out/cxx-qt-gen/qobject.cxx.h:296:68: エラー: static assertion failed: disagrees with the value in #[cxx::bridge]

  cargo:warning=  296 | static_assert(static_cast<::std::uint8_t>(Orientation::Horizontal) == 0, "disagrees with the value in #[cxx::bridge]");

  cargo:warning=      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~

  cargo:warning=/home/username/projects/projectname/target/debug/build/projectname-99c3881ccd5a0cce/out/cxx-qt-gen/qobject.cxx.h:296:68: 備考: the comparison reduces to ‘(1 == 0)’

  cargo:warning=/home/username/projects/projectname/target/debug/build/projectname-99c3881ccd5a0cce/out/cxx-qt-gen/qobject.cxx.h:297:66: エラー: static assertion failed: disagrees with the value in #[cxx::bridge]

  cargo:warning=  297 | static_assert(static_cast<::std::uint8_t>(Orientation::Vertical) == 1, "disagrees with the value in #[cxx::bridge]");

  cargo:warning=      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~

  cargo:warning=/home/username/projects/projectname/target/debug/build/projectname-99c3881ccd5a0cce/out/cxx-qt-gen/qobject.cxx.h:297:66: 備考: the comparison reduces to ‘(2 == 1)’
```


The two below were resolved by modifying the enum on the Rust side.

```
    #[namespace = "Qt"]
    enum Orientation {
        Horizontal = 0x1,
        Vertical = 0x2,
    }
```

The top one, on the other hand, I do not know at this time.
Please help me if you can.

**Reply by ahayzen-kdab** [reply, marked as answer]
You'll need the enum to match the size in Qt, eg if it's a `std::int32_t` in C++ then you need to add `#[repr(i32)]` as an attribute about the enum.

So the following might work if it's a `std::int32_t`

```rust
    #[namespace = "Qt"]
    #[repr(i32)]
    enum Orientation {
        Horizontal = 0x1,
        Vertical = 0x2,
    }
```

**Reply by ghost** [reply]
It works!!
Thanks for all the help.
